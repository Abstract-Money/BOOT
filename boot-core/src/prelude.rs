pub use crate::contract::Contract;
#[cfg(feature = "daemon")]
pub use crate::daemon::{
    core::{instantiate_daemon_env, Daemon},
    state::DaemonOptionsBuilder,
    tx_resp::CosmTxResponse,
};
pub use crate::index_response::IndexResponse;
pub use crate::interface::{
    BootExecute, BootInstantiate, BootMigrate, BootQuery, BootUpload, CallAs, ContractInstance,
};
pub use crate::mock::{
    core::instantiate_custom_mock_env, core::instantiate_default_mock_env, core::Mock,
};
pub use crate::{BootEnvironment, BootError, TxResponse};
pub use boot_contract_derive::boot_contract;
pub use boot_fns_derive::{ExecuteFns, QueryFns};