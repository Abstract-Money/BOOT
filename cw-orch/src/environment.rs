use crate::{state::ChainState, prelude::{IndexResponse, Uploadable}, error::CwOrcError};
use cosmwasm_std::{Addr, BlockInfo, Coin};
use serde::{de::DeserializeOwned, Serialize};
use std::fmt::Debug;

/// Response type for actions on an environment
pub type TxResponse<Chain> = <Chain as TxHandler>::Response;

/// Signals a supported execution environment for CosmWasm contracts
pub trait CwEnv: TxHandler + ChainUpload + Clone {}
impl<T: TxHandler + ChainUpload + Clone> CwEnv for T {}

/// Signer trait for chains.
/// Accesses the sender information from the chain object to perform actions.
pub trait TxHandler: ChainState + Clone {
    type Response: IndexResponse + Debug;
    type Error: Into<CwOrcError> + Debug;
    type ContractSource;

    // Gets current sender
    fn sender(&self) -> Addr;
    // Skip x amount of blocks
    fn wait_blocks(&self, amount: u64) -> Result<(), Self::Error>;
    fn wait_seconds(&self, secs: u64) -> Result<(), Self::Error>;
    fn next_block(&self) -> Result<(), Self::Error>;
    fn block_info(&self) -> Result<BlockInfo, Self::Error>;
    // Actions //
    fn execute<E: Serialize + Debug>(
        &self,
        exec_msg: &E,
        coins: &[Coin],
        contract_address: &Addr,
    ) -> Result<Self::Response, Self::Error>;
    fn instantiate<I: Serialize + Debug>(
        &self,
        code_id: u64,
        init_msg: &I,
        label: Option<&str>,
        admin: Option<&Addr>,
        coins: &[cosmwasm_std::Coin],
    ) -> Result<Self::Response, Self::Error>;
    fn query<Q: Serialize + Debug, T: Serialize + DeserializeOwned>(
        &self,
        query_msg: &Q,
        contract_address: &Addr,
    ) -> Result<T, Self::Error>;
    fn migrate<M: Serialize + Debug>(
        &self,
        migrate_msg: &M,
        new_code_id: u64,
        contract_address: &Addr,
    ) -> Result<Self::Response, Self::Error>;
}

// Required to be a different trait because it can not be implemented for the generic Mock<...>.
pub trait ChainUpload: TxHandler {
    fn upload(&self, contract_source: &impl Uploadable) -> Result<Self::Response, Self::Error>;
}
