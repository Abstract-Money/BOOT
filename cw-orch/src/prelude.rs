//! Glob (*) import me to get all the types you need to get started with cw-orch.
//!
//! ```rust
//! use cw_orch::prelude::*;
//! ```

// We don't want to re-export everything here!
// 1. Macros
// 2. Traits that need to be imported to preform actions on an interface
// 3. Traits that need to be imported to implement an interface
// 4. Objects that need to be available to implement required traits

// Contract traits
pub use crate::contract::interface_traits::{
    CallAs, ConditionalMigrate, ConditionalUpload, ContractInstance, CwOrchExecute,
    CwOrchInstantiate, CwOrchMigrate, CwOrchQuery, CwOrchUpload, ExecutableContract,
    InstantiableContract, MigratableContract, QueryableContract, Uploadable,
};

pub use crate::deploy::Deploy;

pub use crate::environment::StateInterface;

// Response trait
pub use crate::environment::IndexResponse;

// Environment
pub use crate::environment::{CwEnv, QueryHandler, TxHandler, TxResponse};
pub use cw_orch_core::environment::queriers::*;

// Mock for testing
pub use crate::mock::Mock;

// OsmosisTestTube for testing
#[cfg(feature = "osmosis-test-tube")]
pub use crate::osmosis_test_tube::OsmosisTestTube;

// error
pub use crate::error::CwOrchError;

// Paths for implementing `Uploadable`
pub use crate::contract::{ArtifactsDir, WasmPath};

// re-export as it is used in the public API
pub use cosmwasm_std::{Addr, Coin, Empty};
pub use cw_multi_test::{Contract as MockContract, ContractWrapper};

// builder, core type, networks mod, queriers mod, traits
#[cfg(feature = "daemon")]
pub use crate::daemon::{
    live_mock,
    queriers,
    // sync helpers
    Daemon,
    DaemonAsync,
    DaemonAsyncBuilder,
    // expose the sync variants
    DaemonBuilder,
};

#[cfg(feature = "daemon")]
pub use cw_orch_networks::networks;

pub use crate::contract::artifacts_dir_from_workspace;

pub use cw_orch_traits::*;

#[cfg(feature = "snapshot-testing")]
pub use crate::take_storage_snapshot;
