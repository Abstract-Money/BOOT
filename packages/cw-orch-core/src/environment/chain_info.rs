use std::fmt::Display;

use serde::{Deserialize, Serialize};

pub type ChainInfo = ChainInfoBase<&'static str, &'static [&'static str]>;
pub type ChainInfoOwned = ChainInfoBase<String, Vec<String>>;

pub type NetworkInfo = NetworkInfoBase<&'static str>;
pub type NetworkInfoOwned = NetworkInfoBase<String>;

/// Information about a chain.
/// This is used to connect to a chain and to generate transactions.
#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
#[serde(default)]
pub struct ChainInfoBase<StringType: Into<String> + Default, StringArrayType: AsRef<[StringType]>> {
    /// Identifier for the network ex. phoenix-2, pisco-1
    pub chain_id: StringType,
    /// Max gas and denom info
    pub gas_denom: StringType,
    /// gas price
    pub gas_price: f64,
    /// gRPC urls, used to attempt connection
    pub grpc_urls: StringArrayType,
    /// Optional urls for custom functionality
    pub lcd_url: Option<StringType>,
    /// Optional urls for custom functionality
    pub fcd_url: Option<StringType>,
    /// Underlying network details (coin type, address prefix, etc)
    pub network_info: NetworkInfoBase<StringType>,
    /// Chain kind, (local, testnet, mainnet)
    pub kind: ChainKind,
}

/// Information about the underlying network, used for key derivation
#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
#[serde(default)]
pub struct NetworkInfoBase<StringType: Into<String>> {
    /// network identifier (ex. juno, terra2, osmosis, etc)
    pub chain_name: StringType,
    /// address prefix
    pub pub_address_prefix: StringType,
    /// coin type for key derivation
    pub coin_type: u32,
}

impl<StringType: Into<String> + Default> Default for NetworkInfoBase<StringType> {
    fn default() -> Self {
        Self {
            chain_name: StringType::default(),
            pub_address_prefix: StringType::default(),
            // Default cosmos coin
            coin_type: 118,
        }
    }
}

impl<StringType: Into<String> + Default, StringArrayType: AsRef<[StringType]> + Default> Default
    for ChainInfoBase<StringType, StringArrayType>
{
    fn default() -> Self {
        Self {
            chain_id: Default::default(),
            gas_denom: Default::default(),
            gas_price: f64::NAN,
            grpc_urls: Default::default(),
            lcd_url: Default::default(),
            fcd_url: Default::default(),
            network_info: Default::default(),
            kind: Default::default(),
        }
    }
}

impl From<ChainInfo> for ChainInfoOwned {
    fn from(value: ChainInfo) -> Self {
        ChainInfoOwned {
            chain_id: value.chain_id.to_string(),
            gas_denom: value.gas_denom.to_string(),
            gas_price: value.gas_price,
            grpc_urls: value.grpc_urls.iter().map(|url| url.to_string()).collect(),
            lcd_url: value.lcd_url.map(ToString::to_string),
            fcd_url: value.fcd_url.map(ToString::to_string),
            network_info: value.network_info.into(),
            kind: value.kind,
        }
    }
}
impl From<NetworkInfo> for NetworkInfoOwned {
    fn from(value: NetworkInfo) -> Self {
        NetworkInfoOwned {
            chain_name: value.chain_name.to_string(),
            pub_address_prefix: value.pub_address_prefix.to_string(),
            coin_type: value.coin_type,
        }
    }
}

/// Kind of chain (local, testnet, mainnet)
#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ChainKind {
    /// A local chain, used for development
    Local,
    /// A mainnet chain
    Mainnet,
    /// A testnet chain
    Testnet,
    #[default]
    /// Unspecified chain kind
    Unspecified,
}

impl Display for ChainKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let str = match *self {
            ChainKind::Local => "local",
            ChainKind::Testnet => "testnet",
            ChainKind::Mainnet => "mainnet",
            ChainKind::Unspecified => "unspecified",
        };
        write!(f, "{}", str)
    }
}

impl From<String> for ChainKind {
    fn from(str: String) -> Self {
        match str.as_str() {
            "local" => ChainKind::Local,
            "testnet" => ChainKind::Testnet,
            "mainnet" => ChainKind::Mainnet,
            _ => ChainKind::Unspecified,
        }
    }
}

impl<StringType: Into<String> + Default, StringArrayType: AsRef<[StringType]> + Default>
    ChainInfoBase<StringType, StringArrayType>
{
    pub fn config(chain_id: StringType) -> Self {
        Self {
            chain_id,
            ..Default::default()
        }
    }
}

impl ChainInfoOwned {
    /// Overwrite the chain info with the provided chain info.
    pub fn overwrite_with(mut self, chain_info: ChainInfoOwned) -> ChainInfoOwned {
        let ChainInfoOwned {
            chain_id,
            gas_denom,
            gas_price,
            grpc_urls,
            lcd_url,
            fcd_url,
            network_info:
                NetworkInfoOwned {
                    chain_name,
                    pub_address_prefix,
                    coin_type,
                },
            kind,
        } = chain_info;

        if !chain_id.is_empty() {
            self.chain_id = chain_id;
        }
        if !gas_denom.is_empty() {
            self.gas_denom = gas_denom;
        }
        if !gas_price.is_nan() {
            self.gas_price = gas_price;
        }
        if !grpc_urls.is_empty() {
            self.grpc_urls = grpc_urls;
        }
        if let Some(lcd_url) = lcd_url {
            self.lcd_url = Some(lcd_url);
        }
        if let Some(fcd_url) = fcd_url {
            self.fcd_url = Some(fcd_url);
        }
        if !chain_name.is_empty() {
            self.network_info.chain_name = chain_name;
        }
        if !pub_address_prefix.is_empty() {
            self.network_info.pub_address_prefix = pub_address_prefix;
        }
        if coin_type != 118 {
            self.network_info.coin_type = coin_type;
        }
        if kind != ChainKind::Unspecified {
            self.kind = kind;
        }
        self
    }
}
