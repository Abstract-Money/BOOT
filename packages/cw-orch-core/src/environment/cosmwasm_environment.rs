//! Transactional traits for execution environments.

use super::{queriers::QueryHandler, ChainState, IndexResponse};
use crate::{
    contract::interface_traits::{ContractInstance, Uploadable},
    error::CwEnvError,
};
use cosmwasm_std::{Addr, BlockInfo, Coin, ContractInfoResponse};
use serde::{de::DeserializeOwned, Serialize};
use std::fmt::Debug;

/// Signals a supported execution environment for CosmWasm contracts
pub trait CwEnv: TxHandler + QueryHandler + BankQuerier + WasmCodeQuerier + Clone {}
impl<T: TxHandler + QueryHandler + BankQuerier + WasmCodeQuerier + Clone> CwEnv for T {}

/// Response type for actions on an environment
pub type TxResponse<Chain> = <Chain as TxHandler>::Response;

/// Signer trait for chains.
/// Accesses the sender information from the chain object to perform actions.
pub trait TxHandler: ChainState + Clone {
    /// Response type for transactions on an environment.
    type Response: IndexResponse + Debug + Send + Clone;
    /// Error type for transactions on an environment.
    type Error: Into<CwEnvError> + Debug + std::error::Error;
    /// Source type for uploading to the environment.
    type ContractSource;

    type Sender: Clone;

    /// Gets the address of the current wallet used to sign transactions.
    fn sender(&self) -> Addr;

    /// Sets wallet to sign transactions.
    fn set_sender(&mut self, sender: Self::Sender);

    #[deprecated = "Please use `AllQueriers::wait_blocks` instead"]
    /// Wait for an amount of blocks.
    fn wait_blocks(&self, amount: u64) -> Result<(), Self::Error>;

    #[deprecated = "Please use `AllQueriers::wait_seconds` from the NodeQueryHandlerGetter trait instead"]
    /// Wait for an amount of seconds.
    fn wait_seconds(&self, secs: u64) -> Result<(), Self::Error>;

    #[deprecated = "Please use `AllQueriers::next_block` from the NodeQueryHandlerGetter trait instead"]
    /// Wait for next block.
    fn next_block(&self) -> Result<(), Self::Error>;

    #[deprecated = "Please use `AllQueriers::block_info` from the NodeQueryHandlerGetter trait instead"]
    /// Return current block info see [`BlockInfo`].
    fn block_info(&self) -> Result<BlockInfo, Self::Error>;

    // Actions

    /// Uploads a contract to the chain.
    fn upload(&self, contract_source: &impl Uploadable) -> Result<Self::Response, Self::Error>;

    /// Send a InstantiateMsg to a contract.
    fn instantiate<I: Serialize + Debug>(
        &self,
        code_id: u64,
        init_msg: &I,
        label: Option<&str>,
        admin: Option<&Addr>,
        coins: &[cosmwasm_std::Coin],
    ) -> Result<Self::Response, Self::Error>;

    /// Send a ExecMsg to a contract.
    fn execute<E: Serialize + Debug>(
        &self,
        exec_msg: &E,
        coins: &[Coin],
        contract_address: &Addr,
    ) -> Result<Self::Response, Self::Error>;

    #[deprecated = "Please use `AllQueriers::query` instead"]
    /// Send a QueryMsg to a contract.
    fn query<Q: Serialize + Debug, T: Serialize + DeserializeOwned>(
        &self,
        query_msg: &Q,
        contract_address: &Addr,
    ) -> Result<T, Self::Error>;

    /// Send a MigrateMsg to a contract.
    fn migrate<M: Serialize + Debug>(
        &self,
        migrate_msg: &M,
        new_code_id: u64,
        contract_address: &Addr,
    ) -> Result<Self::Response, Self::Error>;

    /// Clones the chain with a different sender.
    /// Usually used to call a contract as a different sender.
    fn call_as(&self, sender: &<Self as TxHandler>::Sender) -> Self {
        let mut chain = self.clone();
        chain.set_sender(sender.clone());
        chain
    }
}

#[deprecated = "Please use .wasm_querier() from the `WasmQuerierGetter` trait"]
pub trait WasmCodeQuerier: TxHandler + Clone {
    #[deprecated = "Please use .wasm_querier().contract_hash from the `WasmQuerierGetter` trait"]
    /// Returns the checksum of provided code_id
    fn contract_hash(&self, code_id: u64) -> Result<String, <Self as TxHandler>::Error>;
    #[deprecated = "Please use .wasm_querier().contract_info from the `WasmQuerierGetter` trait"]
    /// Returns the code_info structure of the provided contract
    fn contract_info<T: ContractInstance<Self>>(
        &self,
        contract: &T,
    ) -> Result<ContractInfoResponse, <Self as TxHandler>::Error>;

    #[deprecated = "Please use .wasm_querier().local_hash from the `WasmQuerierGetter` trait"]
    /// Returns the checksum of the WASM file if the env supports it. Will re-upload every time if not supported.
    fn local_hash<T: Uploadable + ContractInstance<Self>>(
        &self,
        contract: &T,
    ) -> Result<String, CwEnvError> {
        contract.wasm().checksum()
    }
}

#[deprecated = "Please use .bank_querier() from the `BankQuerierGetter` trait"]
pub trait BankQuerier: TxHandler {
    #[deprecated = "Please use .bank_querier().balance from the `BankQuerierGetter` trait"]
    /// Query the bank balance of a given address
    /// If denom is None, returns all balances
    fn balance(
        &self,
        address: impl Into<String>,
        denom: Option<String>,
    ) -> Result<Vec<Coin>, <Self as TxHandler>::Error>;

    #[deprecated = "Please use .bank_querier().supply_of from the `BankQuerierGetter` trait"]
    /// Query total supply in the bank for a denom
    fn supply_of(&self, denom: impl Into<String>) -> Result<Coin, <Self as TxHandler>::Error>;
}
