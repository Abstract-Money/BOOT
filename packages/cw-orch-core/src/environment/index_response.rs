use cosmwasm_std::{Addr, Binary, Event, StdError, StdResult};
#[cfg(feature = "eth")]
use snailquote::unescape;

const CODE_ID_UPLOAD_EVENT: (&str, &str) = ("store_code", "code_id");
const ADDRESS_INSTANTIATE_EVENT: (&str, &str) = ("instantiate", "_contract_address");

#[cfg(feature = "eth")]
const INJECTIVE_CODE_ID_UPLOAD_EVENT: (&str, &str) =
    ("cosmwasm.wasm.v1.EventCodeStored", "code_id");
#[cfg(feature = "eth")]
const INJECTIVE_ADDRESS_INSTANTIATE_EVENT: (&str, &str) = (
    "cosmwasm.wasm.v1.EventContractInstantiated",
    "contract_address",
);

/// Index data returned by transactions which are applicable to both AppResponse (mock env) and TxResponse (live env)
pub trait IndexResponse {
    /// Get all events in the response.
    fn events(&self) -> Vec<Event>;

    /// Search for an event with given attribute key.
    fn event_attr_value(&self, event_type: &str, attr_key: &str) -> StdResult<String>;

    /// Search for all events with a given attribute key.
    fn event_attr_values(&self, _event_type: &str, _attr_key: &str) -> Vec<String>;

    /// Get the data field of the response.
    fn data(&self) -> Option<Binary>;

    /// Helper to get the contract address of a instantiate response.
    fn instantiated_contract_address(&self) -> StdResult<Addr> {
        if let Ok(code_id) = self
            .event_attr_value(ADDRESS_INSTANTIATE_EVENT.0, ADDRESS_INSTANTIATE_EVENT.1)
            .map(Addr::unchecked)
        {
            Ok(code_id)
        } else {
            // for injective
            #[cfg(not(feature = "eth"))]
            panic!("Injective instantiate event parsing not supported without eth feature");
            #[cfg(feature = "eth")]
            return self
                .event_attr_value(
                    INJECTIVE_ADDRESS_INSTANTIATE_EVENT.0,
                    INJECTIVE_ADDRESS_INSTANTIATE_EVENT.1,
                )
                .map(|s| Addr::unchecked(unescape(&s).unwrap()));
        }
    }

    /// Shortcut to get the code id of a contract of an upload response.
    fn uploaded_code_id(&self) -> StdResult<u64> {
        if let Ok(code_id) = self
            .event_attr_value(CODE_ID_UPLOAD_EVENT.0, CODE_ID_UPLOAD_EVENT.1)
            .map(|s| s.parse().unwrap())
        {
            Ok(code_id)
        } else {
            // for injective
            #[cfg(not(feature = "eth"))]
            panic!("Injective upload event parsing not supported without eth feature");
            #[cfg(feature = "eth")]
            self.event_attr_value(
                INJECTIVE_CODE_ID_UPLOAD_EVENT.0,
                INJECTIVE_CODE_ID_UPLOAD_EVENT.1,
            )
            .map(|s| unescape(&s).unwrap().parse().unwrap())
        }
    }
}

/// Cloned from cosmwasm/cw_multi_test
/// A subset of data returned as a response of a contract entry point,
/// such as `instantiate`, `execute` or `migrate`.
#[derive(Default, Clone, Debug)]
pub struct AppResponse {
    /// Response events.
    pub events: Vec<Event>,
    /// Response data.
    pub data: Option<Binary>,
}

#[cfg(feature = "mock")]
impl From<cw_multi_test::AppResponse> for AppResponse {
    fn from(value: cw_multi_test::AppResponse) -> Self {
        Self {
            events: value.events,
            data: value.data,
        }
    }
}

#[cfg(feature = "mock")]
impl From<AppResponse> for cw_multi_test::AppResponse {
    fn from(value: AppResponse) -> Self {
        Self {
            events: value.events,
            data: value.data,
        }
    }
}

impl IndexResponse for AppResponse {
    fn events(&self) -> Vec<Event> {
        self.events.clone()
    }

    fn data(&self) -> Option<Binary> {
        self.data.clone()
    }

    fn event_attr_value(&self, event_type: &str, attr_key: &str) -> StdResult<String> {
        for event in &self.events {
            if event.ty == event_type {
                for attr in &event.attributes {
                    if attr.key == attr_key {
                        return Ok(attr.value.clone());
                    }
                }
            }
        }
        Err(StdError::generic_err(format!(
            "missing combination (event: {}, attribute: {})",
            event_type, attr_key
        )))
    }

    fn event_attr_values(&self, event_type: &str, attr_key: &str) -> Vec<String> {
        let mut all_results = vec![];

        for event in &self.events {
            if event.ty == event_type {
                for attr in &event.attributes {
                    if attr.key == attr_key {
                        all_results.push(attr.value.clone());
                    }
                }
            }
        }
        all_results
    }
}

#[cfg(test)]
mod index_response_test {
    use cosmwasm_std::{Addr, Event};

    use speculoos::prelude::*;

    use super::{AppResponse, IndexResponse};

    const CONTRACT_ADDRESS: &str =
        "cosmos1fd68ah02gr2y8ze7tm9te7m70zlmc7vjyyhs6xlhsdmqqcjud4dql4wpxr";

    fn test_events(idxres: &dyn IndexResponse) -> anyhow::Result<()> {
        asserting!("events length is 1")
            .that(&idxres.events().len())
            .is_equal_to(2);

        Ok(())
    }

    fn test_data(idxres: &dyn IndexResponse) -> anyhow::Result<()> {
        asserting!("data is None").that(&idxres.data()).is_none();

        Ok(())
    }

    fn test_uploaded_code_id(idxres: &dyn IndexResponse) -> anyhow::Result<()> {
        asserting!("uploaded code_id is 1")
            .that(&idxres.uploaded_code_id()?)
            .is_equal_to(1u64);

        Ok(())
    }

    fn test_instantiated_contract_address(idxres: &dyn IndexResponse) -> anyhow::Result<()> {
        asserting!("instantiated contract_address is ")
            .that(&idxres.instantiated_contract_address()?)
            .is_equal_to(Addr::unchecked(CONTRACT_ADDRESS));

        Ok(())
    }

    #[test]
    fn general() {
        let idxres = AppResponse {
            events: vec![
                Event::new("store_code").add_attribute("code_id", "1"),
                Event::new("instantiate")
                    .add_attribute("_contract_address", CONTRACT_ADDRESS.to_owned()),
            ],
            data: None,
        };

        asserting!("test_events is ok")
            .that(&test_events(&idxres))
            .is_ok();

        asserting!("test_data is ok")
            .that(&test_data(&idxres))
            .is_ok();

        asserting!("test_instantiated_contract_address is ok")
            .that(&test_instantiated_contract_address(&idxres))
            .is_ok();

        asserting!("test_uploaded_code_id is ok")
            .that(&test_uploaded_code_id(&idxres))
            .is_ok();
    }
}
