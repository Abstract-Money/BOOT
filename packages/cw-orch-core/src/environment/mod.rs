mod cosmwasm_environment;
mod index_response;
mod mut_env;
pub mod queriers;
mod state;

pub use cosmwasm_environment::{BankQuerier, CwEnv, TxHandler, TxResponse, WasmCodeQuerier};
pub use index_response::IndexResponse;
pub use mut_env::{BankSetter, MutCwEnv};
pub use queriers::QueryHandler;
pub use state::{ChainState, DeployDetails, StateInterface};
