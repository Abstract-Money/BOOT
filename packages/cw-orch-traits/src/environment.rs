use cw_orch_core::{contract::interface_traits::ContractInstance, environment::CwEnv};

pub trait Environment<Chain: CwEnv> {
    // Required method
    fn environment(&self) -> Chain;
}

impl<Chain: CwEnv, T: ContractInstance<Chain>> Environment<Chain> for T {
    fn environment(&self) -> Chain {
        self.get_chain().clone()
    }
}
