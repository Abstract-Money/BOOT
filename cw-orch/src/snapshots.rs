//! Defined the snapshot testing macro on the Mock environment
//! This is included here and not in the mock package because it needs to import traits to work

/// Function helper used to parse storage into readable strings
pub fn parse_storage(storage: &[(Vec<u8>, Vec<u8>)]) -> Vec<(String, String)> {
    storage
        .iter()
        .map(|(key, value)| {
            (
                String::from_utf8_lossy(key).to_string(),
                String::from_utf8_lossy(value).to_string(),
            )
        })
        .collect::<Vec<_>>()
}
/// This functions allows for making sure the tests are stabilised and changes made to contracts don't have an impact on the internal storage
/// This should ONLY be used when cw_orch is in scope
/// Usage:
/// ```rust,ignore
/// take_storage_snapshot!(chain, "mock_doc");
/// ```
/// The name you input to the function should be different from all other snapshots in your repository
/// Find more details on how snapshot testing works on the official quick-start guide: https://insta.rs/docs/quickstart/
/// This function will panic if the snapshot is different from the reference snapshot

#[macro_export]
macro_rules! take_storage_snapshot {
    ($chain: ident, $name: literal) => {
        // We register and test a snapshot for all contracts storage
        use ::cw_orch::environment::{ChainState as _, StateInterface as _};
        let all_contract_addresses = $chain.state().get_all_addresses()?;
        let mut all_storage = vec![];

        for (id, contract_addr) in all_contract_addresses {
            all_storage.push((
                id,
                ::cw_orch::snapshots::parse_storage(
                    &$chain.app.borrow().dump_wasm_raw(&contract_addr),
                ),
            ));
        }
        all_storage.sort_by(|(id_a, _), (id_b, _)| id_a.cmp(id_b));
        ::cw_orch::insta::assert_yaml_snapshot!(
            ::cw_orch::sanitize_filename::sanitize(format!("{}", $name)),
            all_storage
        )
    };
}

#[cfg(test)]
pub mod tests {
    use cosmwasm_std::Addr;
    use cw_orch::prelude::{CwOrchInstantiate, CwOrchUpload, Mock};

    #[test]
    fn contract_snapshots() -> anyhow::Result<()> {
        use counter_contract::CounterExecuteMsgFns;
        let sender = Addr::unchecked("sender");
        let chain = Mock::new(&sender);

        let contract = counter_contract::CounterContract::new("counter_contract", chain.clone());
        contract.upload()?;
        contract.instantiate(
            &counter_contract::msg::InstantiateMsg { count: 0 },
            None,
            None,
        )?;

        contract.increment()?;

        take_storage_snapshot!(chain, "snapshot_test");

        Ok(())
    }
}
