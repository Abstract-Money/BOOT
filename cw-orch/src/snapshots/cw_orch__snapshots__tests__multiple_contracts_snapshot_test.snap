---
source: cw-orch/src/snapshots.rs
expression: all_storage
---
counter_contract:
  - - contract_info
    - "{\"contract\":\"crates.io:counter\",\"version\":\"0.11.0\"}"
  - - state
    - "{\"count\":1,\"owner\":\"sender\"}"
second-counter-contract:
  - - contract_info
    - "{\"contract\":\"crates.io:counter\",\"version\":\"0.11.0\"}"
  - - state
    - "{\"count\":4,\"owner\":\"sender\"}"
third-counter-contract:
  - - contract_info
    - "{\"contract\":\"crates.io:counter\",\"version\":\"0.11.0\"}"
  - - state
    - "{\"count\":3,\"owner\":\"sender\"}"

